engine/language_client_go/callbacks.go

```diff
@@ -0,0 +1,56 @@
+package main
+
+/*
+#cgo LDFLAGS: ./lib/libhello.dylib -ldl
+#include "./lib/baml.h"
+#include <stdlib.h>
+#include <stdbool.h>
+*/
+import "C"
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+)
+
+type ResultCallback struct {
+	data string // JSON string
+}
+
+// Map to store callbacks by ID
+var (
+	dynamicCallbacks = make(map[uint32]chan ResultCallback)
+	callbackMutex    sync.RWMutex
+)
+
+//export trigger_callback
+func trigger_callback(id C.uint32_t, isDone C.bool, result *C.char) {
+	callbackMutex.RLock()
+	id_uint := uint32(id)
+	callback, exists := dynamicCallbacks[id_uint]
+	callbackMutex.RUnlock()
+
+	if exists {
+		my_string := C.GoString(result)
+		fmt.Println("callback triggered", my_string)
+		callback <- ResultCallback{data: my_string}
+		if isDone {
+			close(callback)
+			callbackMutex.Lock()
+			defer callbackMutex.Unlock()
+			delete(dynamicCallbacks, id_uint)
+		}
+	}
+}
+
+func create_unique_id() (C.uint32_t, chan ResultCallback) {
```

Comment:
Why not return a clean go type here? Convert to corresponding C type where it's being consumed